runtime/drivers/athena/athena.go

```diff
@@ -0,0 +1,314 @@
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	prefix := "parquet_output_" + uuid.New().String()
+	bucketName := strings.TrimPrefix(strings.TrimRight(conf.OutputLocation, "/"), "s3://")
+	unloadPath := bucketName + "/" + prefix
+	err = c.unload(ctx, cfg, conf, "s3://"+unloadPath)
+	if err != nil {
+		return nil, errors.Join(fmt.Errorf("failed to unload: %w", err), cleanPath(ctx, cfg, bucketName, prefix))
+	}
+
+	bucketObj, err := c.openBucket(ctx, conf, bucketName)
+	if err != nil {
+		return nil, errors.Join(fmt.Errorf("cannot open bucket %q: %w", unloadPath, err), cleanPath(ctx, cfg, bucketName, prefix))
+	}
+
+	opts := rillblob.Options{
+		GlobPattern: prefix + "/**",
+	}
+
+	it, err := rillblob.NewIterator(ctx, bucketObj, opts, c.logger)
+	if err != nil {
+		return nil, errors.Join(fmt.Errorf("cannot download parquet output %q %w", opts.GlobPattern, err), cleanPath(ctx, cfg, bucketName, prefix))
+	}
+
+	return it, nil
+}
+
+func (c *Connection) openBucket(ctx context.Context, conf *sourceProperties, bucket string) (*blob.Bucket, error) {
+	cfg, err := awsconfig.LoadDefaultConfig(
+		ctx,
+		awsconfig.WithRegion(conf.Region),
+		awsconfig.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(c.config.AccessKeyID, c.config.SecretAccessKey, c.config.SessionToken)),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	s3client := s3v2.NewFromConfig(cfg)
+	return s3blob.OpenBucketV2(ctx, s3client, bucket, nil)
+}
+
+func (c *Connection) unload(ctx context.Context, cfg aws.Config, conf *sourceProperties, path string) error {
+	finalSQL := fmt.Sprintf("UNLOAD (%s) TO '%s' WITH (format = 'PARQUET')", conf.SQL, path)
+	client := athena.NewFromConfig(cfg)
+	resultConfig := &types.ResultConfiguration{
+		OutputLocation: aws.String("s3://" + strings.TrimPrefix(strings.TrimRight(conf.OutputLocation, "/"), "s3://") + "/output/"),
+	}
+
+	executeParams := &athena.StartQueryExecutionInput{
+		QueryString:         aws.String(finalSQL),
+		ResultConfiguration: resultConfig,
+	}
+
+	athenaExecution, err := client.StartQueryExecution(ctx, executeParams)
+	if err != nil {
+		return err
+	}
+
+	r := retrier.New(retrier.ConstantBackoff(20, 1*time.Second), nil)
+
+	return r.RunCtx(ctx, func(ctx context.Context) error {
+		status, err := client.GetQueryExecution(ctx, &athena.GetQueryExecutionInput{
+			QueryExecutionId: athenaExecution.QueryExecutionId,
+		})
+		if err != nil {
+			return err
+		}
+
+		state := status.QueryExecution.Status.State
+
+		if state == types.QueryExecutionStateSucceeded || state == types.QueryExecutionStateCancelled {
+			return nil
+		} else if state == types.QueryExecutionStateFailed {
+			return fmt.Errorf("Athena query execution failed %s", *status.QueryExecution.Status.AthenaError.ErrorMessage)
+		}
+		return fmt.Errorf("Execution is not completed yet, current state: %s", state)
+	})
```

Comment:
1. This seems to translate to a 20 second query execution timeout? It's too little, it should probably continue to check until the `ctx` is cancelled. (It would also return a weird retry error message when hitting the 20s timeout.)
2. It seems it will continue to retry even if query execution failed?
3. I don't see why this warrants a third party library dependency instead of a for loop and simple `select` that checks a timer and context cancellation

Reply:
3. We've already imported that lib and it provides an opinionated approach that one doesn't need to reinvent and it's presents itself right away `retrier`, `ConstantBackoff` - the intention is more recognisable from those words. If I see a low level Timer then I need read more scrupulously to figure out if it's just a simple retry pattern or something else and more complex.